src/viam/sdk/common/proto_t.hpp

```diff
@@ -0,0 +1,293 @@
+// The "simple" value types are empty, bool, int, double, and string.
+// Moreover, a ProtoT can be a vector or string-map of ProtoT.
+class ProtoT {
+   public:
+    // Construct a null object.
+    ProtoT() noexcept : ProtoT(nullptr) {}
+
+    // Construct a nonempty object.
+    template <typename T>
+    ProtoT(T t) : vtable_{model_t<T>::vtable}, self_{std::move(t)} {}
+
+    // Deduction helper constructor for string from string literal
+    ProtoT(const char* str) : ProtoT(std::string(str)) {}
+
+    // Move construction this from other, leaving other in the "unspecified but valid state" of its
+    // moved-from type
+    ProtoT(ProtoT&& other) noexcept
+        : vtable_(std::move(other.vtable_)), self_(std::move(other.self_), vtable_) {}
+
+    ProtoT(const ProtoT& other) : vtable_(other.vtable_), self_(other.self_, other.vtable_) {}
+
+    // Move assignment from other, leaving other in the "unspecified but valid state" of its
+    // moved-from type.
+    ProtoT& operator=(ProtoT&& other) noexcept {
+        ProtoT(std::move(other)).swap(*this);
+        return *this;
+    }
+
+    ProtoT& operator=(const ProtoT& other) {
+        ProtoT(other).swap(*this);
+        return *this;
+    }
+
+    ~ProtoT() {
+        self_.destruct(vtable_);
+    }
+
+    // Test equality of two types.
+    // Note that "intuitive" arithmetic equality is not supported, but could be.
+    // Thus, bool{false}, int{0}, and double{0.0} do not compare equal.
+    friend bool operator==(const ProtoT& lhs, const ProtoT& rhs) {
+        return lhs.vtable_.equal_to(lhs.self_.get(), rhs.self_.get(), rhs.vtable_);
+    }
+
+    void swap(ProtoT& other) {
+        self_.swap(vtable_, other.self_, other.vtable_);
+        std::swap(vtable_, other.vtable_);
+    }
+
+    // Construct from proto value
+    // This method is trivially templated to insulate Value from our API/ABI.
+    // In a translation unit which includes <google/protobuf/struct.pb.h> you can call
+    // this function on a Value instance without specifying a template parameter and it will "just
+    // work"
+    template <typename Value = google::protobuf::Value>
+    static ProtoT from_proto_value(const Value& v) {
+        return ProtoT(&v);
+    }
+
+    // Convert to protobuf Value.
+    friend void to_proto_value(const ProtoT& t, google::protobuf::Value* v);
+
+    // Obtain integer constant representing the stored data type.
+    int kind() const {
+        return vtable_.kind();
+    }
+
+    // Checks whether this ProtoT is an instance of type T.
+    template <typename T>
+    bool is_a() const;
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T* dyn_cast(ProtoT&);
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoT&);
+
+   private:
+    struct vtable_t {
```

Comment:
Do the trailing `_t`'s actually serve a purpose, or could the names be simplified to `vtable`, `storage`, etc.

Reply:
They can be simplified!